deepchem/models/torch_models/grover.py

```diff
@@ -1,94 +1,397 @@
+from deepchem.models.torch_models.grover_layers import (
+    GroverEmbedding, GroverBondVocabPredictor, GroverAtomVocabPredictor,
+    GroverFunctionalGroupPredictor)
+from deepchem.models.torch_models.readout import GroverReadout
 
+
+def _grover_batch_mol(mol_graphs):
+    atom_fdim = 151
+    bond_fdim = 14 + atom_fdim
+
+    n_atoms = 1  # number of atoms (start at 1 b/c need index 0 as padding)
+    n_bonds = 1  # number of bonds (start at 1 b/c need index 0 as padding)
+    a_scope = [
+    ]  # list of tuples indicating (start_atom_index, num_atoms) for each molecule
+    b_scope = [
+    ]  # list of tuples indicating (start_bond_index, num_bonds) for each molecule
+
+    # All start with zero padding so that indexing with zero padding returns zeros
+    f_atoms = [[0] * atom_fdim]  # atom features
+    f_bonds = [[0] * bond_fdim]  # combined atom/bond features
+    a2b = [[]]  # mapping from atom index to incoming bond indices
+    b2a = [
+        0
+    ]  # mapping from bond index to the index of the atom the bond is coming from
+    b2revb = [0]  # mapping from bond index to the index of the reverse bond
+    fg_labels = []  # storing functional group labels
+    for mol_graph in mol_graphs:
+        f_atoms.extend(mol_graph.node_features)
+        f_bonds.extend(mol_graph.edge_features)
+        fg_labels.append(mol_graph.fg_labels)
+        for a in range(mol_graph.num_nodes):
+            a2b.append([b + n_bonds for b in mol_graph.a2b[a]])
+
+        for b in range(mol_graph.num_edges):
+            b2a.append(n_atoms + mol_graph.b2a[b])
+            b2revb.append(n_bonds + mol_graph.b2revb[b])
+
+        a_scope.append((n_atoms, mol_graph.num_nodes))
+        b_scope.append((n_bonds, mol_graph.num_edges))
+        n_atoms += mol_graph.num_nodes
+        n_bonds += mol_graph.num_edges
+
+    # max with 1 to fix a crash in rare case of all single-heavy-atom mols
+    max_num_bonds = max(1, max(len(in_bonds) for in_bonds in a2b))
+
+    f_atoms = torch.FloatTensor(f_atoms)
+    f_bonds = torch.FloatTensor(f_bonds)
+    fg_labels = torch.FloatTensor(np.array(fg_labels))
+    a2b = torch.LongTensor(
+        [a2b[a] + [0] * (max_num_bonds - len(a2b[a])) for a in range(n_atoms)])
+    b2a = torch.LongTensor(b2a)
+    b2revb = torch.LongTensor(b2revb)
+    a2a = b2a[a2b]  # only needed if using atom messages
+    a_scope = torch.LongTensor(a_scope)
+    b_scope = torch.LongTensor(b_scope)
+    return f_atoms, f_bonds, a2b, b2a, b2revb, a_scope, b_scope, a2a
+
+
+class Grover(ModularTorchModel):
+
+    def __init__(self, task='pretraining', model_dir=None):
+        self.task = task
+        self.model_dir = model_dir
+        self.model = self.build_model()
+        self.components = self.build_components()
+        self.classification = mode == "classification"
+        super().__init__(self.model, self.components, **kwargs)
+        self.loss_fn = self.loss_func()
+
+    def build_components(self):
```

Comment:
The component names for the model should be fully documented in the docstring for the model